vavr/src/test/java/org/jdbi/v3/vavr/TestVavrOptionMapperWithDB.java

```diff
@@ -0,0 +1,110 @@
+import org.jdbi.v3.core.rule.H2DatabaseRule;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.sql.SQLException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestVavrOptionMapperWithDB {
+
+    @Rule
+    public H2DatabaseRule dbRule = new H2DatabaseRule().withPlugins();
+
+    @Before
+    public void addData() {
+        Handle handle = dbRule.openHandle();
+        handle.createUpdate("insert into something (id, name) values (1, 'eric')").execute();
+        handle.createUpdate("insert into something (id) values (2)").execute();
+    }
+
+    @Test
+    public void testOptionMapped_shouldSucceed() throws SQLException {
+        final Set<Option<String>> result = dbRule.getSharedHandle()
+                .registerRowMapper(ConstructorMapper.factory(SomethingWithOption.class))
+                .createQuery("select name from something")
+                .collectInto(new GenericType<Set<Option<String>>>() {});
+
+        assertThat(result).hasSize(2);
+        assertThat(result).contains(Option.none(), Option.of("eric"));
+    }
+
+    @Test
+    public void testOptionMappedWithinObjectIfPresent_shouldContainValue() throws SQLException {
+        final SomethingWithOption result = dbRule.getSharedHandle()
+                .registerRowMapper(ConstructorMapper.factory(SomethingWithOption.class))
+                .createQuery("select id, name from something where id = 1")
+                .mapTo(SomethingWithOption.class)
+                .findOnly();
+
+        assertThat(result.getName()).isInstanceOf(Option.class);
+        assertThat(result.getName().get()).isEqualToIgnoringCase("eric");
+    }
+
+    @Test
+    public void testOptionWithinObjectIfMissing_shouldBeNone() throws SQLException {
+        final SomethingWithOption result = dbRule.getSharedHandle()
+                .registerRowMapper(ConstructorMapper.factory(SomethingWithOption.class))
+                .createQuery("select id, name from something where id = 2")
+                .mapTo(SomethingWithOption.class)
+                .findOnly();
+
+        assertThat(result.getName()).isInstanceOf(Option.class);
+        assertThat(result.getName().getOrNull()).isNull();
+    }
+
+    public static class SomethingWithOption {
+        private int id;
+        private Option<String> name;
+
+        public SomethingWithOption(int id, Option<String> name) {
+            this.id = id;
+            this.name = name;
+        }
+
+        public Option<String> getName() {
+            return name;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            SomethingWithOption that = (SomethingWithOption) o;
+
+            if (id != that.id) return false;
+            return name != null ? name.equals(that.name) : that.name == null;
```

Comment:
Aside: I'm assuming this was generated. What IDE are you using? Most of them should be able to generated Java 7-style equals that use `java.util.Objects.equals(Object, Object)` to avoid these null checks.

Reply:
Yeah can redo that - I rarely generate them via IDE since I use lombok

Reply:
Fun fact: Lombok's `@AllArgsConstructor` generates `@java.beans.ConstructorProperties` annotations on the constructor, so you can use `ConstructorMapper` on those types.

Reply:
I know ;) This is the killer feature for me to work with vavr jackson and jdbi for a nice functional rest api